File: libgraffiti/src/layout/style.rs
// TODO: I think this could be an enum, with variant structs for
//       each kind (BlockStyle, FlexStyle, ...) but this is easier for now
//       and maybe it shouldn't be called Style either
//       in OOP there would be a LayoutNode base class and subclasses for
//       each of the node types with their own properties but we don't have
//       a separate tree and we don't have subclasses either

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct LayoutStyle {
    pub display: Display,

    pub size: Size<Dimension>,
    pub min_size: Size<Dimension>,
    pub max_size: Size<Dimension>,

    pub padding: Rect<Dimension>,
    pub margin: Rect<Dimension>,
    pub border: Rect<f32>,

    pub flex_direction: FlexDirection,
    pub flex_wrap: FlexWrap,

    pub flex_grow: f32,
    pub flex_shrink: f32,
    pub flex_basis: Dimension,

    pub align_self: Align,
    pub align_content: Align,
    pub align_items: Align,
    pub justify_content: Justify,

    pub position_type: Position,
    pub position: Rect<Dimension>,
}

impl Default for LayoutStyle {
    fn default() -> Self {
        Self {
            display: Display::Inline,

            size: Size::AUTO,
            min_size: Size::AUTO,
            max_size: Size::AUTO,

            padding: Rect::ZERO,
            margin: Rect::ZERO,
            border: Rect::uniform(0.),

            flex_direction: FlexDirection::Row,
            flex_wrap: FlexWrap::NoWrap,

            flex_grow: 0.,
            flex_shrink: 1.,
            flex_basis: Dimension::Auto,

            align_self: Align::Auto,
            align_items: Align::Stretch,
            align_content: Align::Stretch,
            justify_content: Justify::FlexStart,

            position_type: Position::Static,
            position: Rect::AUTO,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Display {
    None,
    Block,
    Inline,
    InlineBlock,
    Flex,
    Table,
    TableRow,
    TableCell,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Dimension {
    Auto,
    Px(f32),
    Percent(f32),
    Em(f32),
    Rem(f32),
    Vw(f32),
    Vh(f32),
    Vmin(f32),
    Vmax(f32),
}

#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct Size<T: Copy> {
    pub width: T,
    pub height: T,
}

impl<T: Copy> Size<T> {
    pub const fn new(width: T, height: T) -> Self {
        Self { width, height }
    }

    pub const fn uniform(a: T) -> Self {
        Self::new(a, a)
    }
}

impl<T: Copy + PartialOrd> Size<T> {
    pub fn min(self) -> T {
        if self.width < self.height {
            self.width
        } else {
            self.height
        }
    }

    pub fn max(self) -> T {
        if self.width > self.height {
            self.width
        } else {
            self.height
        }
    }
}

impl Size<Dimension> {
    pub const AUTO: Self = Self::uniform(Dimension::Auto);
    pub const ZERO: Self = Self::uniform(Dimension::Px(0.));
}

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub struct Rect<T: Copy> {
    pub top: T,
    pub right: T,
    pub bottom: T,
    pub left: T,
}

impl<T: Copy> Rect<T> {
    pub const fn new(top: T, right: T, bottom: T, left: T) -> Self {
        Self {
            top,
            right,
            bottom,
            left,
        }
    }

    pub const fn uniform(v: T) -> Self {
        Self::new(v, v, v, v)
    }
}

impl Rect<Dimension> {
    pub const AUTO: Self = Self::uniform(Dimension::Auto);
    pub const ZERO: Self = Self::uniform(Dimension::Px(0.));
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum FlexDirection {
    Row,
    Column,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum FlexWrap {
    NoWrap,
    Wrap,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Align {
    Auto,
    FlexStart,
    Center,
    FlexEnd,
    Stretch,
    Baseline,
    SpaceBetween,
    SpaceAround,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Justify {
    FlexStart,
    Center,
    FlexEnd,
    SpaceBetween,
    SpaceAround,
    SpaceEvenly,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Position {
    Static,
    Relative,
    Absolute,
}

File: src/layout/style.zig
const std = @import("std");

// longhand because it's easier to map from CSS
pub const LayoutStyle = struct {
    display: Display = .block,

    width: Dimension = Dimension.auto,
    height: Dimension = Dimension.auto,
    // min_width: Dimension = Dimension.auto,
    // min_height: Dimension = Dimension.auto,
    // max_width: Dimension = Dimension.auto,
    // max_height: Dimension = Dimension.auto,

    padding_top: Dimension = Dimension.ZERO,
    padding_right: Dimension = Dimension.ZERO,
    padding_bottom: Dimension = Dimension.ZERO,
    padding_left: Dimension = Dimension.ZERO,

    // margin_top: Dimension = Dimension.ZERO,
    // margin_right: Dimension = Dimension.ZERO,
    // margin_bottom: Dimension = Dimension.ZERO,
    // margin_left: Dimension = Dimension.ZERO,

    // border_top: Dimension = Dimension.ZERO,
    // border_right: Dimension = Dimension.ZERO,
    // border_bottom: Dimension = Dimension.ZERO,
    // border_left: Dimension = Dimension.ZERO,

    // TODO: position, top, right, left

    flex_grow: f32 = 0,
    flex_shrink: f32 = 1,
    flex_basis: Dimension = Dimension.auto,

    flex_direction: FlexDirection = .row,
    flex_wrap: FlexWrap = .no_wrap,

    align_content: AlignContent = .stretch,
    align_items: AlignItems = .stretch,
    align_self: AlignSelf = .auto,
    justify_content: JustifyContent = .flex_start,
};

pub const Display = enum {
    none,
    block,
    flex,
    @"inline",
};

pub const FlexDirection = enum {
    row,
    column,
    row_reverse,
    column_reverse,
};

pub const FlexWrap = enum {
    no_wrap,
    wrap,
    wrap_reverse,
};

pub const AlignContent = enum {
    flex_start,
    center,
    flex_end,
    stretch,
    space_between,
    space_around,
    space_evenly,
};

pub const AlignItems = enum {
    flex_start,
    center,
    flex_end,
    baseline,
    stretch,
};

pub const AlignSelf = enum {
    auto,
    flex_start,
    center,
    flex_end,
    baseline,
    stretch,
};

pub const JustifyContent = enum {
    flex_start,
    center,
    flex_end,
    space_between,
    space_around,
    space_evenly,
};

pub const Dimension = union(enum) {
    auto,
    px: f32,
    percent: f32,

    const Self = @This();

    pub const ZERO = Self{ .px = 0 };

    pub fn resolve(self: Self, base: f32) f32 {
        return switch (self) {
            .auto => std.math.nan_f32,
            .px => |v| v,
            .percent => |v| v / 100 * base,
        };
    }
};
